Screens: Add settings screen

kristofbolyai opened the pull request.

Incompleteusern commented on src/main/java/com/wynntils/modules/utilities/overlays/inventories/FavoriteItemsOverlay.java:
```diff
@@ -0,0 +1,103 @@
+            if (nbt.hasKey("wynntilsFavorite")) continue; // already been checked
+
+            String itemName = StringUtils.normalizeBadString(getTextWithoutFormattingCodes(stack.getDisplayName()));
+            if (!itemName.contains("Unidentified")) {
```
I don't think this should be inverted

HeyZeer0 commented on src/main/java/com/wynntils/modules/utilities/overlays/inventories/FavoriteItemsOverlay.java:
```diff
@@ -0,0 +1,103 @@
+            ScreenRenderer.beginGL(0, 0);
+            GlStateManager.translate(0, 0, 300);
+
+            ScreenRenderer r = new ScreenRenderer();
```
This will cause some heavy GC, don't we have a static instance of this stored somewhere? If not just create one on this class and it should be fine.

HeyZeer0 commented on src/main/java/com/wynntils/modules/utilities/overlays/inventories/FavoriteItemsOverlay.java:
```diff
@@ -0,0 +1,103 @@
+
+            ScreenRenderer r = new ScreenRenderer();
+            RenderHelper.disableStandardItemLighting();
+            float scaleFactor = 0.5f;
+            ScreenRenderer.scale(scaleFactor);
```
Declaring the variable here seems unnecessary? Probably could be
```suggestion
            ScreenRenderer.scale(0.5f);
```

HeyZeer0 commented on src/main/java/com/wynntils/modules/utilities/overlays/inventories/FavoriteItemsOverlay.java:
```diff
@@ -0,0 +1,103 @@
+            RenderHelper.disableStandardItemLighting();
+            float scaleFactor = 0.5f;
+            ScreenRenderer.scale(scaleFactor);
+            r.drawRect(Textures.Map.map_icons, (int)((gui.getGuiLeft() + s.xPos + 10) / scaleFactor), (int)((gui.getGuiTop() + s.yPos - 5) / scaleFactor), 208, 36, 18, 18);
```
Instead of translating into 0, 0 with beginGl, translate to `gui.getGuiLeft()` and `gui.getGuiTop()` including the offset to the center of the slot, that will make you not need to divide by the scale factor reducing overhead and improving alias.

HeyZeer0 commented on src/main/java/com/wynntils/modules/utilities/overlays/inventories/FavoriteItemsOverlay.java:
```diff
@@ -0,0 +1,103 @@
+                }
+            }
+
+            if (!nbt.hasKey("wynntilsFavorite")) // wasn't favorited
+                nbt.setBoolean("wynntilsFavorite", false);
```
Any reason why you are setting it to false instead of checking if it contains? This will probably add a bunch of unnecessary data but might have a reason

P0keDev replied:
I did it so that any item only needs to be checked once, and then can be ignored on every tick afterwards - I figured that would be more efficient

HeyZeer0 commented on src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java:
```diff
@@ -659,6 +666,11 @@ public void handleAccessoryMovement(TickEvent.ClientTickEvent e) {
 
     @SubscribeEvent
     public void clickOnChest(GuiOverlapEvent.ChestOverlap.HandleMouseClick e) {
+        if (e.getSlotIn() != null && e.getSlotId() - e.getGui().getLowerInv().getSizeInventory() == 4) { // prevent accidental pouch clicking
+            e.setCanceled(true);
+            return;
+        }
```
Does this block clicks on the pouch with the pouch open? If so might wanna check if that doesn't cause any weird behavior

P0keDev replied:
Clicking the pouch while in the pouch menu force closes it in the same way as any other chest gui, so disabling it shouldn't be an issue

HeyZeer0 reviewed (changes requested):
LGTM once changes were made. Really nice work on this! Definitely a cool feature and thanks for the bug fixes 😄

kristofbolyai committed:
Screens: Refactor module structure

kristofbolyai committed:
RenderUtils: Use floats

kristofbolyai committed:
SettingsScreen: Basic layout

kristofbolyai committed:
WynntilsSettingsScreen: Features are now displayed

kristofbolyai committed:
WynntilsSettingsScreen: Fix search

kristofbolyai commented:
Status update 2022.08.22: 
Feature list is now semi-working with basic search functionality. 

![image](https://user-images.githubusercontent.com/49001742/186003698-4b2620b8-e69f-4690-bf29-78080e074215.png)
![image](https://user-images.githubusercontent.com/49001742/186003805-a49b6acf-9a0a-43e5-a3e4-c01fde36078b.png)

kristofbolyai committed:
Merge branch 'main' into settings_screen

kristofbolyai committed:
I18n: Add missing feature translations

Incompleteusern commented:
I feel that rounded edges of the button kinda clashes as the scroll bar is already round, on that note I think making it look similar to overlay config (red if enabled, green if disabled, square boxes) might work. In general, it might be good to make feature config and overlay config look more similar (though I definitely prefer the feature config color choices).

In the future, maybe we could have customizable color palettes for the settings screen, but that's a bit meta and might have problems, along with being infeasible right now.

magicus commented:
I think we should not sweat to much over graphic design at this point. Hopefully we can get some graphic designers to step up; my impression is that there is usually a ratio of 3:1 of people wanting to contribute who can do graphics relative to those who can do coding.

kristofbolyai commented:
Also I would like to mention the pain that working with MC screens/guis cause. It is, for sure, something we need to tackle when we get the new textures. If there is no good framework we can shadow, we might as well create one...

magicus commented:
There are GUI frameworks, of course. For configuration screens, for instance, there are several popular, ClothConfig for instance. I'm not sure how well they will fit the "look and feel" of Wynncraft, though. Or if we think that matters.

kristofbolyai committed:
FeatureList: Refactor + fix scrolling bug

kristofbolyai committed:
SearchWidget: Add a basic cursor to the text box

kristofbolyai commented:
Status update 2022.08.24: 
Today was focused on small bug fixes in the feature list code. Now scrolling is less buggy, and mostly works (and looks) fine. I also tested different resolutions, like 1080p, and the screen is looking okay (there are some alignment issues with multi-line feature names, we can fix that by taking out "feature" of the end of every feature name).

I also went ahead and improved on our search widget, as I believe it will be something we will re-use in many different places. Now it has a working cursor!
![image](https://user-images.githubusercontent.com/49001742/186527420-75a7dc9d-8b81-41ff-9c09-cefd96c45115.png)

Incompleteusern commented on common/src/main/java/com/wynntils/screens/settings/lists/FeatureList.java:
```diff
@@ -113,8 +104,6 @@ protected void renderBackground(PoseStack poseStack) {
     }
 
-    // FIXME: This is incorrect, entries have different heights
-    //        Consider not using this in renderList
```
How is this fixed now?

kristofbolyai replied:
It is not, but it does not make that big of a difference. I will fix it either way though (although not here, but in the render code where I actually use this).

kristofbolyai replied:
Fixed now.

magicus commented:
I agree that we can take out "Feature". It looked good in the beginning, but in the context of all features it just looks weird.

magicus commented:
Regarding the search (or rather "filter") box, one feature I tend to like for such is the ability to match not only substrings, but just consecutive characters. Like "heapob" would match "HEAlth POtion Blocker". (It could be matched by "loo" as well, "heaLth pOtiOn blocker", so if we do that, we should probably reorder results so those where the match *is* consecutive("LOOtrun") comes on top.)

Not saying you have to do this, but if you get inspired and think it sounds like a cool thing to implement, I won't protest. :)

kristofbolyai committed:
FeatureList: Make list smoother

kristofbolyai committed:
SearchWidget: Add much better search matching

kristofbolyai committed:
I18n: Remove "Feature" from feature names

kristofbolyai commented:
Status update 2022.08.25:
More improvements to feature list code and search bar.

Localization now does not have " Feature" for every feature name.

![image](https://user-images.githubusercontent.com/49001742/186754425-7bd3a407-1122-4649-96d5-16adcd4f1d83.png)

Search bar now supports "fuzzy"(ish) matching (basically what @magicus described https://github.com/Wynntils/Artemis/pull/345#issuecomment-1226926441).

![image](https://user-images.githubusercontent.com/49001742/186754456-25142409-dbb9-472e-a5a2-1c93212689da.png)
![image](https://user-images.githubusercontent.com/49001742/186754480-8329a68c-4cc2-42ef-bf00-0f060d63b2f7.png)

kristofbolyai committed:
FeatureList: Add a search "button"

kristofbolyai commented:
Status update 2022.08.25 part 2: 

Scroll button!

![image](https://user-images.githubusercontent.com/49001742/186775618-43b3c49a-5914-4f42-a005-7b3ab12bbb45.png)


PS: We also hit 1k lines of code... I haven't even added anything about selecting a feature, and displaying it's specific config options :sweat_smile:

kristofbolyai committed:
FeatureList: Do not show debug features

HighCrit commented on common/src/main/java/com/wynntils/screens/settings/lists/FeatureList.java:
```diff
@@ -0,0 +1,262 @@
+        String lastCategory = "";
+
+        for (Feature feature : FeatureRegistry.getFeatures().stream()
+                .filter(feature -> !(feature instanceof DebugFeature))
```
```suggestion
                .filter(feature -> !(feature instanceof DebugFeature) || WynntilsMod.isDevelopmentEnvironment())
```
I think it'd be nice for our devs to easily access debugfeatures in menu.

kristofbolyai replied:
Maybe, but I am not sure if we should do this. Devs can and should be able to use the config command and/or edit the config files, then reload the configs with already existing command.

magicus replied:
I agree. Or we can hand-edit the files to enable/disable features and recompile.

kristofbolyai replied:
> I agree. Or we can hand-edit the files to enable/disable features and recompile.

Can you clarify which comment you agree with 😅?

Incompleteusern replied:
I agree that it would still be more convenient to use the menu

magicus replied:
I agree with the agreeableness of all agreements. Agreed? :-)

Seriously, though, I thought that debug features do not need GUI configs. They are a special case tool. If we start adding them in devel builds, it only means the config screen will be not-as-in-production for devel builds, which is bad, for very little gain.

kristofbolyai committed:
WynntilsSettingsScreen: Add base FeatureSettingWidget (rendering for rounded rects)

HighCrit commented on common/src/main/java/com/wynntils/mc/render/RenderUtils.java:
```diff
@@ -390,6 +390,98 @@ public static void drawArc(
+
+        bufferBuilder.end();
+        BufferUploader.end(bufferBuilder);
+        RenderSystem.disableBlend();
```
Why don't you just call drawArc and calculate the values of the arguments in drawRoundedCorner

kristofbolyai replied:
Done. Had to modify drawArc a bit though... I first thought PoseStack translations would work, but they did not.

kristofbolyai committed:
RenderUtils: Simplify drawRoundedCorner

kristofbolyai committed:
FeatureSettingWidget: Add colors and padding

kristofbolyai committed:
FeatureList: Fix vertical alignment for categories

kristofbolyai committed:
Merge branch 'main' into settings_screen

kristofbolyai commented:
Status update 2022.08.27: 

Vertically aligned categories.

![image](https://user-images.githubusercontent.com/49001742/187046644-db9b520c-df99-415a-bb44-c058e9a45e93.png)

kristofbolyai committed:
FeatureList: Add the ability to select and display selected feature

kristofbolyai committed:
FeatureList + FeatureSettingWidget: Click and sound improvements

kristofbolyai committed:
FeatureSettingWidget: Render scrollbar

kristofbolyai commented:
Status update 2022.08.27 part 2:

The settings screen buttons were updated. Now there is the expected close without saving, and save buttons. 

Features can be selected now. Selection can be moved up/down with arrows. Selecting a feature will play the UI click sound. 

Selected feature name is now displayed, and a "scrollbar" is drawn. This is in preparation for config list rendering.

![image](https://user-images.githubusercontent.com/49001742/187050607-dda326ee-1d5f-4536-a20e-b54b7cf78958.png)

Incompleteusern commented:
It might be nice to color the boxes instead later on, but currently this is not supported and has no need to go out of the way to support.

kristofbolyai committed:
FeatureList: Reduce code duplication

kristofbolyai commented:
@Incompleteusern https://github.com/Wynntils/Artemis/pull/345/commits/eb9522b158b8cae24d6b0ac4191402ab6a6a3ced solves code duplication, but it has a little cost...
I can't return early + I have to use atomic references.

Incompleteusern commented:
> @Incompleteusern [eb9522b](https://github.com/Wynntils/Artemis/commit/eb9522b158b8cae24d6b0ac4191402ab6a6a3ced) solves code duplication, but it has a little cost... I can't return early + I have to use atomic references.

Feel free to revert that if the cost is too much in your opinion, both are fine for me.

kristofbolyai commented:
> > @Incompleteusern [eb9522b](https://github.com/Wynntils/Artemis/commit/eb9522b158b8cae24d6b0ac4191402ab6a6a3ced) solves code duplication, but it has a little cost... I can't return early + I have to use atomic references.
> 
> Feel free to revert that if the cost is too much in your opinion, both are fine for me.

I think it is fine until I only update hovered on click. If we have to do it in render, to render a tooltip or smth, I will definitely revert.

kristofbolyai committed:
WynntilsSettingsScreen: Make BAR_HEIGHT static

kristofbolyai committed:
FontRenderer: Fix one alignment bug in renderAlignedTextInBox

kristofbolyai committed:
FeatureSettingWidget: Add basic config option rendering

kristofbolyai committed:
Merge branch 'main' into settings_screen

kristofbolyai committed:
FeatureSettingWidget: Add scroll bar

kristofbolyai committed:
FeatureSettingWidget: Render descriptions

kristofbolyai committed:
Merge branch 'main' into settings_screen

kristofbolyai committed:
chore: spotless formatting

kristofbolyai committed:
Merge branch 'main' into settings_screen

kristofbolyai committed:
FeatureList: Fix merge

kristofbolyai committed:
Merge branch 'main' into settings_screen

kristofbolyai committed:
Merge branch 'main' into settings_screen

Incompleteusern commented on common/src/main/java/com/wynntils/screens/settings/widgets/FeatureSettingWidget.java:
```diff
@@ -0,0 +1,189 @@
+            ConfigOptionElement configWidgetRight = configWidgets.size() > i + 1 ? configWidgets.get(i + 1) : null;
+            float renderHeight = this.height / 4f;
+
+            if (renderY + renderHeight > this.y + this.height) break;
```
This can be simplified to
```java
            if (renderY > this.y + 3/4f * this.height) break;
```

kristofbolyai replied:
I am not sure this refactor is much simpler either.

Incompleteusern replied:
It gets rid of the `renderHeight` variable

kristofbolyai replied:
Well I need the renderHeight variable now anyways

kristofbolyai committed:
FeatureSettingWidget: Add on/off switch for feature

kristofbolyai committed:
FeatureSettingWidget: Add on/off for boolean configs

kristofbolyai committed:
WynntilsSettingsScreen: Esc is now exit without saving

kristofbolyai commented:
Status update 2022.08.28:

Features now display their config options, with scroll bar!
Feature userEnabled state can be toggled with a switch.
Boolean config options can be toggled.
Setting screen ESC now reverts settings to their previous state.
Fixed FontRenderer alignment bug.

![image](https://user-images.githubusercontent.com/49001742/187097811-c0aa213a-921a-4f3e-9fa6-5c4337a67e37.png)

magicus commented:
So what about my suggestion to have the enable switch next to the actual feature, in the list? I'm thinking that would make it very easy to see if a feature is actually enabled, and to turn it on and off.

kristofbolyai commented:
> So what about my suggestion to have the enable switch next to the actual feature, in the list? I'm thinking that would make it very easy to see if a feature is actually enabled, and to turn it on and off.

I think I missed it, if you mentioned this before. I can do that. Fine by me either way.

kristofbolyai committed:
Merge branch 'main' into settings_screen

magicus commented:
I tried to describe it in the discord channel, but without a mockup or sketch it's hard to get across ideas of GUI design. 

I think it would make for a very smooth UX, so if you can do it, please do.

Incompleteusern commented on common/src/main/java/com/wynntils/screens/settings/elements/BooleanConfigOptionElement.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Override
+    protected void renderConfigAppropriateButton(PoseStack poseStack, float width, float height) {
+        if (configHolder.getType().equals(Boolean.class)) {
```
Isn't this unnecessary?

kristofbolyai replied:
It is now. There was a need in another iterations.

Incompleteusern commented on common/src/main/java/com/wynntils/screens/settings/widgets/FeatureSettingWidget.java:
```diff
@@ -0,0 +1,308 @@
+    private Feature cachedFeature = null;
+    private List<ConfigOptionElement> configWidgets = new ArrayList<>();
+    private ConfigOptionElement hoveredConfigWidget = null;
+    private boolean enabledStateChangeable = false;
```
This variable is unused as of now

kristofbolyai replied:
It is not? I check it when rendering the on off switch?

kristofbolyai replied:
I will get rid of it anyways if I move the switches.

Incompleteusern replied:
You opted to use
```java
         final Texture switchTexture = selectedFeature.isEnabled() ? Texture.SWITCH_ON : Texture.SWITCH_OFF;
```
there instead

kristofbolyai commented on common/src/main/java/com/wynntils/screens/settings/widgets/FeatureSettingWidget.java:
```diff
@@ -0,0 +1,308 @@
+    }
+
+    private void renderEnabledSwitch(PoseStack poseStack, Feature selectedFeature) {
+        if (!enabledStateChangeable) return;
```
@Incompleteusern

Incompleteusern replied:
ohhhhh

kristofbolyai committed:
Merge branch 'main' into settings_screen

kristofbolyai commented:
Status update 2022.08.29:

Text box inputs! I generalized the search widget, and now we have a text input widget as well. Now we properly handle and care about what text box is selected/focused, compared to when we only had the search widget.

Text config types: They have a parse indicator. I am planning for these to work with most config holders, and we can add special support for numbers, etc. later.

I also changed the size of the switches. They look better like this on smaller displays.

This screen now reached a state where it is completely usable to edit any feature configs. I only have to add overlay config editing, and it will be "ready" to ship. However, I do plan to add a special button for enums, so it is easier to switch between them.

![image](https://user-images.githubusercontent.com/49001742/187294077-ec561701-b6b2-4489-a643-a7ab19b394a5.png)
![image](https://user-images.githubusercontent.com/49001742/187295041-cac0ab9e-d79f-4783-a89b-1d2b31ca7fae.png)

![image](https://user-images.githubusercontent.com/49001742/187294248-8379b4e2-1449-4b5f-b8f5-2774ff720a80.png)

kristofbolyai committed:
FeatureSettingWidget: Add text boxes

kristofbolyai commented:
Also, almost hit 2000 lines of code.

Incompleteusern commented on common/src/main/java/com/wynntils/screens/settings/elements/BooleanConfigOptionElement.java:
```diff
@@ -16,41 +16,39 @@ public BooleanConfigOptionElement(ConfigHolder configHolder, FeatureSettingWidge
         if (value == null) {
             configHolder.setValue(false);
         } else {
             configHolder.setValue(!value);
```
This sets the value, so even if you quit without saving, the config is still applied when `Configmanager#save` is called

kristofbolyai replied:
Yeah but we reload saved configs from the file. So it is either saved or discarded.

kristofbolyai committed:
Merge branch 'main' into settings_screen

kristofbolyai committed:
WynntilsSettingsScreen: Move feature enabled switches to FeatureList/FeatureEntry

kristofbolyai commented:
![image](https://user-images.githubusercontent.com/49001742/187450335-0ce7a0a3-22a8-4b0d-b530-84ed3956782f.png)

kristofbolyai committed:
FeatureSettingWidget: Add EnumConfigOptionElement

kristofbolyai committed:
FeatureSettingWidget: Add CustomColorConfigOptionElement

kristofbolyai commented:
Status update 2022.08.30:

This is finally ready for reviews!

Changes since last time:
Moved the feature switch
Added custom config switches for enums and custom colors
Fixed stuff

Tested on 1080p and 3440*1440. This is likely not perfect on every resolution, but we can deal with strange cases later.

Screenshots of the current state:
![image](https://user-images.githubusercontent.com/49001742/187468893-68446488-14c8-42ba-b674-beaf580447c6.png)
![image](https://user-images.githubusercontent.com/49001742/187468906-1d99c36e-2c58-438d-a95c-273c4e89abd8.png)
![image](https://user-images.githubusercontent.com/49001742/187468925-730709ae-934c-4d26-978e-75f1bf5aa146.png)
![image](https://user-images.githubusercontent.com/49001742/187468992-7b125a1c-f81a-4c72-82ce-eb4266729d99.png)

kristofbolyai commented:
Let's start the review process. This is gonna take a few iterations :\).

Incompleteusern commented on common/src/main/java/com/wynntils/screens/widgets/SearchWidget.java:
```diff
@@ -0,0 +1,98 @@
+                        this.width,
+                        CommonColors.WHITE,
+                        FontRenderer.TextAlignment.LEFT_ALIGNED,
+                        FontRenderer.TextShadow.NORMAL);
```
Cursor char should render when focused, even if empty

kristofbolyai replied:
Why? When we are not focused, we render "Search...", when we are focused, we render the current search with the cursor char.

Incompleteusern replied:
Last time I checked, when we are focused and empty, it still renders "Search..."

kristofbolyai replied:
![image](https://user-images.githubusercontent.com/49001742/187482601-d3a4631f-8607-4ab0-8802-09c551509b36.png)
![image](https://user-images.githubusercontent.com/49001742/187482629-22abc6a9-1ad0-40a0-9e59-f88725c18c0b.png)

kristofbolyai replied:
![image](https://user-images.githubusercontent.com/49001742/187482727-28b27a77-6df6-4223-99b2-6dabfd8d4bd5.png)

kristofbolyai replied:
I am not sure when you checked. It is working now, I might have fixed it yesterday or today though.

Incompleteusern commented on common/src/main/java/com/wynntils/screens/widgets/SearchWidget.java:
```diff
@@ -0,0 +1,98 @@
+                Texture.SEARCH_BAR.width(),
+                Texture.SEARCH_BAR.height(),
+                Texture.SEARCH_BAR.width(),
+                Texture.SEARCH_BAR.height());
```
Don't we have a method already to render an entire bar? If not, we should have one.

kristofbolyai replied:
The method we already have is not scaling up the texture, but rather, rendering it "multiple" times as a pattern.

kristofbolyai replied:
Fixed.

Incompleteusern commented on common/src/main/java/com/wynntils/screens/widgets/TextInputBoxWidget.java:
```diff
@@ -0,0 +1,215 @@
+            }
+
+            renderedText = builder.toString();
+        }
```
What does this code do?

kristofbolyai replied:
This is basically to only render text that fits into the text widget (truncate if string is too long). I can add code comments to document this.

kristofbolyai committed:
Merge branch 'main' into settings_screen

magicus commented:
I plan to start by checking out the PR and testing it locally. I understand that at some point, behavioral fixes might have to wait for follow up PRs, but I want to see that it works roughly okay for me.

kristofbolyai committed:
chore: spotless formatting

kristofbolyai committed:
TextInputBoxWidget: Add comment

kristofbolyai committed:
RenderUtils: Add drawScalingTexturedRect

kristofbolyai committed:
Merge branch 'main' into settings_screen

Incompleteusern committed:
Merge branch 'main' into settings_screen

Incompleteusern commented:
<img width="570" alt="image" src="https://user-images.githubusercontent.com/58920010/187594617-17a04844-8b15-4023-891e-5f93ef898a5e.png">
The text boxes are small for my screen size, but that is best for a separate pr.

On another note, the addition and removal of the cursor is annoying, as it shifts the text back and forth, but is not that big of a concern.

A bug occurs when the cursor is placed at the end of a text input book that has enough characters to fill it entirely, since then the cursor position can desync when pressing the left key once when its rendered.

This is caused by
```java
            int stringPosition = Math.min(textBoxInput.length() - 1, cursorPosition);
```
where cursorPosition equals `textBoxInput.length()`; subtracting one causes no change in rendering but changes the cursor position;

A fix is likely to replace
```java
        cursorPosition = Math.min(textBoxInput.length(), cursorPosition + 1);
```
with
```java
        cursorPosition = Math.min(textBoxInput.length() - 1, cursorPosition + 1);
```
which on a note also makes the earlier `Math.min` for `stringPosition` moot, if that is correct.

kristofbolyai commented:
> <img alt="image" width="570" src="https://user-images.githubusercontent.com/58920010/187594617-17a04844-8b15-4023-891e-5f93ef898a5e.png">
> 
> The text boxes are small for my screen size, but that is best for a separate pr.
> On another note, the addition and removal of the cursor is annoying, as it shifts the text back and forth, but is not that big of a concern.
> 
> A bug occurs when the cursor is placed at the end of a text input book that has enough characters to fill it entirely, since then the cursor position can desync when pressing the left key once when its rendered.
> 
> This is caused by
> 
> ```java
>             int stringPosition = Math.min(textBoxInput.length() - 1, cursorPosition);
> ```
> 
> where cursorPosition equals `textBoxInput.length()`; subtracting one causes no change in rendering but changes the cursor position;
> 
> A fix is likely to replace
> 
> ```java
>         cursorPosition = Math.min(textBoxInput.length(), cursorPosition + 1);
> ```
> 
> with
> 
> ```java
>         cursorPosition = Math.min(textBoxInput.length() - 1, cursorPosition + 1);
> ```
> 
> which on a note also makes the earlier `Math.min` for `stringPosition` moot, if that is correct.

I will increase the size, but I've tried fixing the blinking and it did not work, if you want, you can work on that as a follow up.

kristofbolyai committed:
TextConfigOptionElement: Increase text box width

kristofbolyai committed:
FeatureSettingWidget: Add overlay config options

kristofbolyai commented:
Added support for overlay config options, I left that out by accident. 

![image](https://user-images.githubusercontent.com/49001742/187641058-bf813591-352b-473e-ba5f-1fd536a6d420.png)

kristofbolyai committed:
Elements: Remove useless class

kristofbolyai committed:
Merge branch 'main' into settings_screen

kristofbolyai committed:
Merge branch 'main' into settings_screen

kristofbolyai committed:
Merge branch 'main' into settings_screen

kristofbolyai committed:
Merge branch 'main' into settings_screen

kristofbolyai committed:
Merge branch 'main' into settings_screen

magicus commented:
So, I need to take the bull by the horns. Making a deep-dive review now.

magicus commented on common/src/main/java/com/wynntils/core/features/Feature.java:
```diff
@@ -210,6 +210,10 @@ public boolean canEnable() {
     }
 
+    public boolean canUserEnable() {
+        return getConfigOptionFromString("userEnabled").isPresent();
```
Seem a bit backwards to determine this from config strings...

kristofbolyai replied:
This could be an instaceof check too, yes.

magicus replied:
I'd prefer that.

magicus commented on common/src/main/java/com/wynntils/mc/render/RenderUtils.java:
```diff
@@ -383,6 +406,68 @@ public static void drawArc(
+        // Corners
+        poseStack.pushPose();
+        poseStack.translate(-1, -1, 0);
+        drawRoundedCorner(poseStack, borderColor, x, y, z, innerRadius, outerRadius, Mth.HALF_PI * 3);
```
`Mth`?

kristofbolyai replied:
Mth is Minecraft's MathUtils class.

Incompleteusern replied:
Name of minecraft math class.

magicus replied:
Ah. Weird name. It used to be MathUtils, iirc. Much better imho.

magicus commented on common/src/main/java/com/wynntils/screens/overlays/OverlaySelectionScreen.java:
```diff
@@ -24,9 +24,11 @@ public class OverlaySelectionScreen extends Screen {
+    private final Screen lastScreen;
 
     public OverlaySelectionScreen() {
         super(new TranslatableComponent("screens.wynntils.overlaySelection.name"));
```
I think in general there is a pattern in Minecraft where you send in the "previous" screen as the first argument to the Screen constructor. There used to be in 1.12 at least. I think we can try and follow it.

kristofbolyai replied:
I will see if there is such thing here. I don't think so though.

kristofbolyai replied:
No such thing in 1.18.2

magicus commented on common/src/main/java/com/wynntils/screens/settings/WynntilsSettingsScreen.java:
```diff
@@ -0,0 +1,239 @@
+    public boolean mouseScrolled(double mouseX, double mouseY, double delta) {
+        if (featureList.isMouseOver(mouseX, mouseY)) {
+            return featureList.mouseScrolled(mouseX, mouseY, delta);
+        } else if (featureSettingWidget.isMouseOver(mouseX, mouseY)) {
```
Is this not handled by super, since `featureSettingWidget` is set by `addRenderableWidget`?

kristofbolyai replied:
This is needed so the "order" of processing the click is how I want it, compared to super.

magicus replied:
I might be wrong here, but if not, that would also make a case for also adding the `featureList` to the renderable widget list. I think that is in general more akin to how the guis of minecraft work. I understand this has been in the works for long, and I will not require any additional works just "because it looks better", but if you agree with me you might consider this :)

kristofbolyai replied:
I don't think feature list can be added as a widget though.

kristofbolyai replied:
(because of the superclass)

magicus replied:
ok

magicus commented on common/src/main/java/com/wynntils/screens/settings/lists/FeatureList.java:
```diff
@@ -0,0 +1,362 @@
+
+    @Override
+    protected FeatureListEntryBase getEntryAtPosition(double mouseX, double mouseY) {
+        AtomicReference<FeatureEntry> returnValue = new AtomicReference<>(null);
```
Why AtomicReference? Is there a race here?

kristofbolyai replied:
Needed for the code below.

magicus commented on common/src/main/java/com/wynntils/screens/widgets/SearchWidget.java:
```diff
@@ -0,0 +1,94 @@
+            int height,
+            Consumer<String> onUpdateConsumer,
+            WynntilsSettingsScreen settingsScreen) {
+        super(x, y, width, height, new TextComponent("Search..."), onUpdateConsumer, settingsScreen);
```
Why not use `DEFAULT_TEXT` here?

kristofbolyai replied:
This is actually the widget name, so it should not be "search.."

magicus commented on common/src/main/java/com/wynntils/screens/widgets/SearchWidget.java:
```diff
@@ -0,0 +1,94 @@
+
+    @Override
+    public void render(PoseStack poseStack, int mouseX, int mouseY, float partialTick) {
+        this.renderBg(poseStack, McUtils.mc(), mouseX, mouseY);
```
Maybe rename it `renderBackground` to match other places?

kristofbolyai replied:
This is what the superclass method is called.

magicus replied:
And you can't use that? What's the difference by the super's backgorund render method, and your?

kristofbolyai replied:
I mean superclass's method is called renderBg, and I would rather override that than make a new method.

magicus replied:
Ah, ok.

magicus commented on common/src/main/java/com/wynntils/screens/widgets/SearchWidget.java:
```diff
@@ -0,0 +1,94 @@
+                        this.x + this.width - 5,
+                        this.y + 6.5f,
+                        this.width,
+                        CommonColors.WHITE,
```
If we are showing the default "Search..." string, make it gray or light gray to show more clearly that it is not a real search term.

magicus commented on common/src/main/java/com/wynntils/screens/widgets/SearchWidget.java:
```diff
@@ -0,0 +1,94 @@
+
+    @Override
+    protected void removeFocus() {
+        this.setTextBoxInput("");
```
What does this do? Remove the search term if you e.g. tab away from it? Is that really correct?

kristofbolyai replied:
Remove search if you esc away:

Incompleteusern replied:
Could this be undesirable?

kristofbolyai replied:
You can click away and search will stay. Esc for quick search reset

magicus commented on common/src/main/java/com/wynntils/screens/widgets/TextInputBoxWidget.java:
```diff
@@ -0,0 +1,216 @@
+
+        poseStack.translate(this.x, this.y, 0);
+
+        this.renderBg(poseStack, McUtils.mc(), mouseX, mouseY);
```
`renderBackground`?

kristofbolyai replied:
Same as before, superclass method name

magicus commented on common/src/main/java/com/wynntils/screens/settings/widgets/FeatureSettingWidget.java:
```diff
@@ -0,0 +1,306 @@
+
+                if (configWidgetRight.getConfigHolder().getParent() instanceof Overlay
+                        && configWidgetRight.getConfigHolder().getParent() != lastOverlayConfigTitle) {
+                    renderY += renderHeight + padding;
```
I'm not sure I understand the if condition here. At what circumstance do we want/need to render the right widget to the left?

kristofbolyai replied:
When there is an overlay, we need to render the title. If the left config field is the first for the overlay, it is easy case. If it is the right, we shift the right config title to be the left one, also rendering the title next iteration.

magicus commented on common/src/main/java/com/wynntils/screens/settings/elements/CustomColorConfigOptionElement.java:
```diff
@@ -0,0 +1,44 @@
+        poseStack.translate(-getTextInputHeight() * 1.2f, 0, 0);
+
+        if (!this.lastParseSuccessful) {
+            RenderUtils.drawLine(
```
Maybe add comment "make a red cross if parsing failed" or something like that. Took me a while to figure out.

Incompleteusern committed:
Merge branch 'main' into settings_screen

Incompleteusern committed:
Fix merge conflicts

Incompleteusern committed:
chore: spotless formatting

magicus commented on common/src/main/java/com/wynntils/screens/settings/elements/ConfigOptionElement.java:
```diff
@@ -9,11 +9,13 @@
 import com.wynntils.screens.settings.WynntilsSettingsScreen;
 import com.wynntils.screens.settings.widgets.FeatureSettingWidget;
 import net.minecraft.client.gui.narration.NarrationElementOutput;
+import oshi.hardware.CentralProcessor;
```
:-D I guess spotless will take care of that.

Incompleteusern committed:
Merge branch 'main' into settings_screen

kristofbolyai committed:
Merge branch 'main' into settings_screen

kristofbolyai commented:
@Incompleteusern text alignment still broken 
![image](https://user-images.githubusercontent.com/49001742/188293134-f173d4e6-5007-45f4-9ebf-7c59da49ea10.png)

kristofbolyai committed:
WynntilsSettingsScreen: Review changes

kristofbolyai committed:
WynntilsSettingsScreen: Fix FontRenderer weirdness

kristofbolyai commented:
Fixed.

magicus committed:
Merge branch 'main' into settings_screen

magicus reviewed (dismissed):
Looks good enough for now.

magicus commented:
We will have to continue tweaking this, though. This is how it looks on my screen with lower resolution. :-D
![image](https://user-images.githubusercontent.com/6209108/188316728-096b5111-4c89-4b5b-9ae7-0c19b451d378.png)

magicus commented:
But it's good enough for now. Let's ship it and move on.

kristofbolyai committed:
ConfigOptionElement: Increase description size

kristofbolyai committed:
EnumConfigOptionElement: Another FontRenderer merge fix
